admin-core/src/main/kotlin/com/epam/drill/admin/endpoints/instance/AgentInstanceEndpoints.kt

```diff
@@ -0,0 +1,187 @@
+import io.ktor.routing.*
+import io.ktor.util.pipeline.*
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.json.Json
+import kotlinx.serialization.protobuf.ProtoBuf
+import kotlinx.serialization.serializer
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.util.zip.GZIPInputStream
+
+private val logger = KotlinLogging.logger {}
+
+@Location("/api/agents")
+object Agents {
+
+    @Location("/{agentId}/builds/{buildVersion}/coverage")
+    data class Coverage(val agentId: String, val buildVersion: String)
+
+    @Location("/{agentId}/builds/{buildVersion}/class-metadata")
+    data class ClassMetadata(val agentId: String, val buildVersion: String)
+
+    @Location("/{agentId}/builds/{buildVersion}/class-metadata/complete")
+    data class ClassMetadataComplete(val agentId: String, val buildVersion: String)
+}
+
+class AgentInstanceEndpoints(override val di: DI) : DIAware {
+    private val app by instance<Application>()
+    private val agentManager by instance<AgentManager>()
+    private val pd by instance<PluginDispatcher>()
+
+    init {
+        app.routing {
+            registerAgentInstanceRoute()
+            sendCoverageRoute()
+            sendClassMetadataRoute()
+            completeSendingClassMetadataRoute()
+        }
+    }
+
+    private fun Route.registerAgentInstanceRoute() {
+        put<Agents> {
+            val agentConfig = call.decompressAndReceive<AgentConfig>()
+            val agentInfo: AgentInfo = withContext(Dispatchers.IO) {
+                agentManager.attach(agentConfig)
+            }
+            processPluginData(agentInfo, InitInfo())
+            call.respond(HttpStatusCode.OK)
+        }
+    }
+
+    private fun Route.sendCoverageRoute() {
+        post<Agents.Coverage> { params ->
+            handleAgentRequest(params.agentId, params.buildVersion) { agentInfo ->
+                val data = call.decompressAndReceive<CoverageData>()
+                processPluginData(agentInfo, data.toCoverDataPart())
+            }
+        }
+    }
+
+    private fun Route.sendClassMetadataRoute() {
+        post<Agents.ClassMetadata> { params ->
+            handleAgentRequest(params.agentId, params.buildVersion) { agentInfo ->
+                val data = call.decompressAndReceive<ClassMetadata>()
+                processPluginData(agentInfo, data.toInitDataPart())
+            }
+        }
+    }
+
+    private fun Route.completeSendingClassMetadataRoute() {
+        post<Agents.ClassMetadataComplete> { params ->
+            handleAgentRequest(params.agentId, params.buildVersion) { agentInfo ->
+                processPluginData(agentInfo, Initialized())
+            }
+        }
+    }
+
+    private suspend fun PipelineContext<Unit, ApplicationCall>.handleAgentRequest(
+        agentId: String,
+        buildVersion: String,
+        handler: suspend (AgentInfo) -> Any
+    ) {
+        agentManager.getOrNull(agentId)
+            ?.let { agentInfo ->
```

Comment:
Prefer flat-ter code structure to nesting / avoid unnecessary nesting.
In this case - `val` + `if` check to `?.let` and elvis op
```
val something = agentManager.getOrNull()
if (something.isNull()) {
  // handle null
}
// proceed with the main flow
```